### PR TITLE
1492 window ventilation

### PR DESCRIPTION
**Pull request by Jun-Jiang-92**
Add the package "WindowVentilation" under "AixLib.Airflow". This package provides several empirical expressions based on the literature to estimate the air volume flow by single-sided window opening.

Link to issue #1492.

### CODE REVIEW

**Comment by FWuellhorst** [comment]
@Jun-Jiang-92 Thanks for the big new package! I would suggest to split up the review, so that others get customed to review, as well. @MZuschlag, are you interested in the review of this?

**Comment by Jun-Jiang-92** [comment]
@FWuellhorst I renewed (almost) all the code in this new package to meet the rules of namespace. At the same time, some models were reconstructed and debugged for a more stable performance.
The pipeline in git-CI doesn't work at the moment. I don't know why.
Would you be so kind as to review my code? For questions I'm available with pleasure.

File: AixLib/Airflow/WindowVentilation/Utilities/WindProfilePowerLaw.mo [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+    annotation (Placement(transformation(extent={{-140,-20},{-100,20}})));
+  Modelica.Blocks.Interfaces.RealOutput winSpe(unit="m/s", min=0)
+    "Wind speed at height"
+    annotation (Placement(transformation(extent={{100,-10},{120,10}})));
+equation
+  assert(
+    height > lenRuf,
+    "Power law not applicable by height less than the roughness length",
+    AssertionLevel.error);
+  cof = 1/log(height/lenRuf);
```
Comment (FWuellhorst) [inline AixLib/Airflow/WindowVentilation/Utilities/WindProfilePowerLaw.mo:21]: Isn't cof a parameter? Both inputs are at least.

File: AixLib/Airflow/WindowVentilation/BaseClasses/PartialOpeningArea.mo [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+  Modelica.Blocks.Interfaces.RealInput opnWidth_in(
+    final quantity="Length", final unit="m", min=0) if use_opnWidth_in
+    "Conditional input port for window sash opening width"
+    annotation (Placement(transformation(extent={{-140,-20},{-100,20}})));
+  Modelica.Blocks.Interfaces.RealOutput A(
+    final quantity="Area", final unit="m2", min=0)
+    "Output port for window opening area, must be defined by extended model"
+    annotation (Placement(transformation(extent={{100,-10},{120,10}})));
+  Modelica.Blocks.Interfaces.RealOutput opnWidth(final unit="m", min=0)
+    "Internal port to connect to opnWidth_in or prescribedOpnWidth";
```
Comment (FWuellhorst) [inline AixLib/Airflow/WindowVentilation/BaseClasses/PartialOpeningArea.mo:26]: We normally use `_internal` for such cases, so the input is `opnWidth` and the internally used variable `opnWidth_internal`

**Review by FWuellhorst — CHANGES_REQUESTED** [review]
@Jun-Jiang-92 The model looks really nice, thanks for your work!
Beside the inline comments, I have some general requests:
1. The examples produce warnings, which should not be the case. The error messages look more like "debug" messages, which may confuse users. Is this message really necessary? 
```
Warning: The following was detected at time: 0
  The polynomial under the square root to calculate V_flow_th is less than 0, the V_flow_th will be set to 0
  Failed condition: warrenParkins.intRes > 1E-15
```
2. You have a lot of assertions, which are fine. But you can move the ones which depend on parameters to "initial equation" or "initial algorithm" section to indicate those are design assertations not triggered during simulation. 
3. You use some if else with `>=`, e.g. on the incidence angle `AixLib.Airflow.WindowVentilation.EmpiricalExpressions.LarsenHeiselberg`. If users control the window opening based on the value and set the set value of the control to e.g. 105 in this case, you will get a lot of state events. So double check if greater equal comparisons are "robust" in the way that the variables are not going to be controlled. Else, don't use the "equal" part or even better, a small hysteresis.
4. Can you add "simulate and plot" scripts for your examples? This way, you ensure the results won't change in future commits or at least you get an error and can check what did change. 
5. Increase the diagram size in the examples to fit all model instances

File: AixLib/Airflow/WindowVentilation/BaseClasses/PartialEmpiricalFlowStack.mo [hunk 3]
```diff
@@ -0,0 +1,28 @@
+﻿within AixLib.Airflow.WindowVentilation.BaseClasses;
+partial model PartialEmpiricalFlowStack
+  "Partial model for empirical expressions with stack effect considered"
+  extends AixLib.Airflow.WindowVentilation.BaseClasses.PartialEmpiricalFlow;
+  Modelica.Blocks.Interfaces.RealInput TRoom(
+    final unit="K", min=273, max=313)
+    "Room temperature, ranging from 0 to 40 °C"
```
[the remark this answers is not in the document]
Comment (Jun-Jiang-92) [reply AixLib/Airflow/WindowVentilation/BaseClasses/PartialEmpiricalFlowStack.mo:7]: This package is primarily developed for HOM. The window geometry must be always exactly defined with clear width, clear height, opening width, and even more parameters. So, generally, it is too detailed for ROM.
If it is fine to keep the `TRoom` then I will keep it. It's more intuitive and coherent with the definition in most literature.

File: AixLib/Airflow/WindowVentilation/BaseClasses/PartialEmpiricalFlowStack.mo [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+    annotation (Placement(transformation(extent={{-140,60},{-100,100}})));
+  Modelica.Blocks.Interfaces.RealInput TAmb(
+    final unit="K", min=243, max=323)
+    "Ambient temperature, ranging from -30 to 50 °C"
+    annotation (Placement(transformation(extent={{-140,20},{-100,60}})));
```
[the remark this answers is not in the document]
Comment (Jun-Jiang-92) [reply AixLib/Airflow/WindowVentilation/BaseClasses/PartialEmpiricalFlowStack.mo:12]: The two values here can be considered as placeholders. Empirical models of window ventilation are often developed with field tests or CFD-simulations. In these cases, data will be acquired with limited boundary conditions, i.e. room and ambient temperature. Therefore, I limited the input here.
Not all the literature mention the valid range of their models. So, I put a dummy range here. This range could be modified for each model (so, I didn't use `final` prefix).

[the remark this answers is not in the document]
Comment (FWuellhorst) [reply AixLib/Airflow/WindowVentilation/BaseClasses/PartialEmpiricalFlowStack.mo:12]: But it is just the doc, not the min/max of the value.

[the remark this answers is not in the document]
Comment (Jun-Jiang-92) [reply AixLib/Airflow/WindowVentilation/BaseClasses/PartialEmpiricalFlowStack.mo:12]: They are alreay there.
https://github.com/RWTH-EBC/AixLib/blob/fde457c2d17395626e76f6546437da40b2ff9dd1/AixLib/Airflow/WindowVentilation/BaseClasses/PartialEmpiricalFlowStack.mo#L10

File: AixLib/Airflow/WindowVentilation/BaseClasses/Types/SmallestAngleDifferenceTypes.mo [hunk 5]
```diff
@@ -0,0 +1,8 @@
+﻿within AixLib.Airflow.WindowVentilation.BaseClasses.Types;
+type SmallestAngleDifferenceTypes = enumeration(
+    Range180 "Result in the range of -180° to +180°",
+    Range360 "Result in the range of 0 to 360°")
+    "Enumeration to define smallest angle difference types for the function"
```
[the remark this answers is not in the document]
Comment (Jun-Jiang-92) [reply AixLib/Airflow/WindowVentilation/BaseClasses/Types/SmallestAngleDifferenceTypes.mo:5]: I linked this enumeration, also the eunumeration `SunshadingIntallationTypesVDI2078` to the functions in DocStr.

File: AixLib/Airflow/WindowVentilation/Utilities/WindProfilePowerLaw.mo (hunk 1 above)
Comment (Jun-Jiang-92) [reply AixLib/Airflow/WindowVentilation/Utilities/WindProfilePowerLaw.mo:21]: Yes, you are right. I redefined it as `final parameter`.

File: AixLib/Airflow/WindowVentilation/BaseClasses/PartialOpeningArea.mo (hunk 2 above)
Comment (Jun-Jiang-92) [reply AixLib/Airflow/WindowVentilation/BaseClasses/PartialOpeningArea.mo:26]: Already renamed.

File: AixLib/Airflow/WindowVentilation/BaseClasses/PartialEmpiricalFlowStack.mo (hunk 4 above)
[the remark this answers is not in the document]
Comment (FWuellhorst) [reply AixLib/Airflow/WindowVentilation/BaseClasses/PartialEmpiricalFlowStack.mo:12]: Ah, missed that!

**Comment by Jun-Jiang-92** [comment]
@FWuellhorst Thank you for your review! I already pushed the revised version. Regarding your questions:

>     1. The examples produce warnings, which should not be the case. The error messages look more like "debug" messages, which may confuse users. Is this message really necessary?
They are not debug-msgs but indeed necessary warning-msgs. A lot of empirical models in the literature do not consider the sign of polynomial under the square root, because this polynomial is always positive with their measurment data (also used as training data), which are conducted with limited boundary conditions that did not cover all the temperauture and wind ranges. By negative polynomial, I replace it to 0, in order to keep the simulation running.
Therefore, these empirical models will become inaccurate if the polynomial is negative, in which case the warning-msg is important.
The examples produce a lot of warnings, because I set a wide range of inputs there. This makes these examples like a "pressure test". But in practice, the warning will be rarely triggerd. I already test these models with some field test data from the facade test bench, there was only few warnings.

>     2. You have a lot of assertions, which are fine. But you can move the ones which depend on parameters to "initial equation" or "initial algorithm" section to indicate those are design assertations not triggered during simulation.
Good point. I already moved assertations depended on params to "initial equation".

>     3. You use some if else with `>=`, e.g. on the incidence angle `AixLib.Airflow.WindowVentilation.EmpiricalExpressions.LarsenHeiselberg`. If users control the window opening based on the value and set the set value of the control to e.g. 105 in this case, you will get a lot of state events. So double check if greater equal comparisons are "robust" in the way that the variables are not going to be controlled. Else, don't use the "equal" part or even better, a small hysteresis.
I changed '>=' to '>'. The incidence angle is the difference of wind direction (from weather data) and the azimut of window / facade (parameter, fixed). The zero crossing will happen by intepolating the weather data. So, there will not be lots of event. The most other comparisons, specially comparing the "polynomial under square root" with 0, are all set with `NoEvent` to avoid zero crossing.

>     4. Can you add "simulate and plot" scripts for your examples? This way, you ensure the results won't change in future commits or at least you get an error and can check what did change.
I added them and hope they work well.

>     5. Increase the diagram size in the examples to fit all model instances
Done.

If you have any questions, feel free to ge in touch with me.

**Comment by FWuellhorst** [comment]
I changed all models except your own expression, as this extends `AixLib.Airflow.WindowVentilation.BaseClasses.PartialEmpiricalFlow`. Please check if usage of extends is possible in this case as well. 
Also, the simulate and plot scripts fail in ci:
```
*** Error: Failed to find Tolerance in '/builds/EBC/EBC_all/github_ci/AixLib/AixLib/../AixLib/Airflow/WindowVentilation/Examples/OpeningArea.mo'.
*** Error: Failed to find Tolerance in '/builds/EBC/EBC_all/github_ci/AixLib/AixLib/../AixLib/Airflow/WindowVentilation/Examples/VentilationFlowRateSimpleOpening.mo'.
*** Error: Failed to find Tolerance in '/builds/EBC/EBC_all/github_ci/AixLib/AixLib/../AixLib/Airflow/WindowVentilation/Examples/VentilationFlowRateSashOpening.mo'.```
Can you add the Tolerance in the experiment setup in the example?

**Comment by Jun-Jiang-92** [comment]
@FWuellhorst I rebuilt the assertion check into `AixLib/Airflow/WindowVentilation/BaseClasses/PartialEmpiricalFlow.mo`, so that every ventilation expression uses this check. But somehow, the warning triggered by t=0 will not be counted by the counter. So for now, the first two warnings will be shown in the simulation examples. Do you know how to fix it?

File: AixLib/Airflow/WindowVentilation/BaseClasses/PartialEmpiricalFlow.mo [hunk 6]
```diff
@@ -28,7 +28,24 @@ partial model PartialEmpiricalFlow
...
 equation
+  // Assertion check
+  when intRes < Modelica.Constants.eps then
+    errCouIntRes = pre(errCouIntRes) + 1;
+  end when;
+  assert(intRes > Modelica.Constants.eps or errCouIntRes > 1,
+    "In " + getInstanceName() + ": The polynomial under the square root to 
+    calculate '" + varNameIntRes + "' is equal or less than 0, '" +
+    varNameIntRes + "' will be set to 0",
+    AssertionLevel.warning);
```
Comment (FWuellhorst) [inline AixLib/Airflow/WindowVentilation/BaseClasses/PartialEmpiricalFlow.mo:47]: I think using `or errCouIntRes` will lead to assertions whenever the error count is bigger than one, or? It should be `and errCouIntRes < 2`?
I don't know how to definitly raise it only once, but maybe the following version could do the trick:
```
  when intRes < Modelica.Constants.eps then
    errCouIntRes = pre(errCouIntRes) + 1;
    if errCouIntRes == 1 then
     assert(true,
    "In " + getInstanceName() + ": The polynomial under the square root to 
    calculate '" + varNameIntRes + "' is equal or less than 0, '" +
    varNameIntRes + "' will be set to 0",
    AssertionLevel.warning);
    end if;
  end when;
 
```

Comment (Jun-Jiang-92) [reply AixLib/Airflow/WindowVentilation/BaseClasses/PartialEmpiricalFlow.mo:47]: Think about this case:
The `intRes` is greater than `eps`, as it should be, this leads `intRes > Modelica.Constants.eps` to be `True`.
The `errCouIntRes` is greater than 2 due to previous warnings, this leads `errCouIntRes < 2` to be `False`.
By using `and`, you get `assert(True and False)`, that is `assert(False)`. That leads to a warning for all normal cases when warnings have been triggered more than once. Is that right?

Comment (FWuellhorst) [reply AixLib/Airflow/WindowVentilation/BaseClasses/PartialEmpiricalFlow.mo:47]: ah yes, I switched the required "false" in the assert in my head. 
I don't think we should try to avoid the duplicate warning, as I am not sure if the order in which the when and assert are triggered will always be the same. Two warnings are better than possibly thousand.

**Comment by FWuellhorst** [comment]
@Jun-Jiang-92 My last comments, as I had a closer look at the examples to implement the assertion:
1. You connect inputs with inputs. This is not recommend, as it's more complex to analyze where an input originates. Also, if you delete a block with an unused output, you would not expect missing inputs in other instances. 
2. The simulation of 180 s generates a whooping 333 state events. Using simulation analysis, the main driver is `larsenHeiselberg` (ca. 200 events), then `caciolo` (ca. 80 events) and then `vdi2078_1` (18 events). Removing these three options yields only 12 events  (which are still many for 180s. The remainder of events are for maas (ca. 8), tang (2) and jiang (one) at weird times (see picture). 
Coupled to a high-order model with multiple windows and zones, this behavior will drastically increase simulation time of the already slow HOM. I am not sure if you can avoid the events or if the approaches are not interesting for you, but in any case a small hint in the doc of these models would be fair for users so they know what causes the slow simulation.

File: AixLib/Airflow/WindowVentilation/BaseClasses/PartialEmpiricalFlow.mo (hunk 6 above)
Comment (Jun-Jiang-92) [reply AixLib/Airflow/WindowVentilation/BaseClasses/PartialEmpiricalFlow.mo:47]: OK, I will keep it as the current version.

**Comment by Jun-Jiang-92** [comment]
@FWuellhorst I reconnected all of the inputs directly to sources. Thank you for reminding.

Regarding the huge amount of events, they are because I set the examples again into 'stress test' with large ranges of input conditions (see the first explanation in https://github.com/RWTH-EBC/AixLib/pull/1495#issuecomment-2220984666). In practice use, this will not happen.

I already added more DocStr into examples to explain it. If you like, I can revert the boudary condition settings back to `f170771` that no warnings will be triggered.

**Review by FWuellhorst — APPROVED** [review]
Not necessary, it looks good! Thank you for this big work, they seem like really cool models :) Let's wait approx 30 min to see if the CI still passes, which it looks like. Then you can merge.
